metadata/src/lib.rs

```diff
@@ -3,844 +3,120 @@
-    let mut bytes = hash(&[19]);
-    let mut visited_ids = HashSet::<u32>::new();
-
-    if let Some(calls) = &pallet.calls {
-        bytes = xor(
-            bytes,
-            get_type_hash(registry, calls.ty.id, &mut visited_ids),
-        );
-    }
-    if let Some(ref event) = pallet.event {
-        bytes = xor(
-            bytes,
-            get_type_hash(registry, event.ty.id, &mut visited_ids),
-        );
-    }
-    for constant in pallet.constants.iter() {
-        bytes = xor(bytes, hash(constant.name.as_bytes()));
-        bytes = xor(
-            bytes,
-            get_type_hash(registry, constant.ty.id, &mut visited_ids),
-        );
-    }
-    if let Some(ref error) = pallet.error {
-        bytes = xor(
-            bytes,
-            get_type_hash(registry, error.ty.id, &mut visited_ids),
-        );
-    }
-    if let Some(ref storage) = pallet.storage {
-        bytes = xor(bytes, hash(storage.prefix.as_bytes()));
-        for entry in storage.entries.iter() {
-            bytes = hash_hashes(
-                bytes,
-                get_storage_entry_hash(registry, entry, &mut visited_ids),
-            );
-        }
-    }
-
-    bytes
-}
-
-/// Obtain the hash representation of a `frame_metadata::RuntimeMetadataV14`.
-pub fn get_metadata_hash(metadata: &RuntimeMetadataV14) -> [u8; 32] {
-    // Collect all pairs of (pallet name, pallet hash).
-    let mut pallets: Vec<(&str, [u8; 32])> = metadata
-        .pallets
-        .iter()
-        .map(|pallet| {
-            let hash = get_pallet_hash(&metadata.types, pallet);
-            (&*pallet.name, hash)
-        })
-        .collect();
-
-    // Sort by pallet name to create a deterministic representation of the underlying metadata.
-    pallets.sort_by_key(|&(name, _hash)| name);
-
-    // Note: pallet name is excluded from hashing.
-    // Each pallet has a hash of 32 bytes, and the vector is extended with
-    // extrinsic hash and metadata ty hash (2 * 32).
-    let mut bytes = Vec::with_capacity(pallets.len() * 32 + 64);
-    for (_, hash) in pallets.iter() {
-        bytes.extend(hash)
-    }
-
-    bytes.extend(get_extrinsic_hash(&metadata.types, &metadata.extrinsic));
 
-    let mut visited_ids = HashSet::<u32>::new();
-    bytes.extend(get_type_hash(
-        &metadata.types,
-        metadata.ty.id,
-        &mut visited_ids,
-    ));
-
-    hash(&bytes)
-}
+pub use retain::retain_metadata_pallets;
+pub use validation::{
+    get_call_hash, get_constant_hash, get_metadata_hash, get_metadata_per_pallet_hash,
+    get_pallet_hash, get_storage_hash, NotFound,
+};
 
-/// Obtain the hash representation of a `frame_metadata::RuntimeMetadataV14`
-/// hashing only the provided pallets.
+/// Latest runtime metadata version supported by subxt.
+pub type LatestRuntimeMetadata = RuntimeMetadataV15;
+
+/// Convert the metadata V14 to the latest metadata version.
+pub fn metadata_v14_to_latest(metadata: RuntimeMetadataV14) -> LatestRuntimeMetadata {
```

Comment:
I'm happy for this in this PR, but ultimately wonder whether we want to make the version of metadata we use in Subxt opaque, eg have a struct like

```
pub struct RuntimeMetadata(RuntimeMetadataV15);
```

and then having the functions in this crate accept this as an argument.

The idea is that then, changes to the metadata (esp as V15 is unstable and liable to change a little for instance) will not be a breaking change to Subxt, just an internal detail.

If we did this, there is a question about whether users may want access to the metadata that Subxt is using though. (Subxt wraps it and ends up with a `Metadata` struct, and currently users can call certain functions on that (the ones we use in Subxt internally). I don't really like that interface being exposed (I'd rather Metadata was entirely opaque) but a compromise is that we allow access to the underlying RuntimeMetadataV14 or whatever.

So ultimately I think we could:

- Have `RuntimeMetadata` as above to hide away the concrete version we are working with
- Make all of the functions on `Metadata` (in subxt lib crate) `pub(crate)`; that way we can have exactly the interfacce we need as Subxt evolves internally without worrying about breaking things for users (because they can't use that stuff
- Give users access to the underlying metadata by `impl RuntimeMetadata { pun fn into_runtime_metadata_v14(self) -> RuntimeMetadataV14 }` and eg `impl Metadata { pub fn runtime_metadata() }` (so that user can do `metadata.runtime_metadata().clone().into_runtime_metadata_v14()`; yes you'd need to clone but it'd be a one off need; maybe we can do better?)

Then, in the future we can add `into_runtime_metadata_v15()` etc and so on as a purely additive non breaking change and the actual metadata we use internally never matters to users (and can be eg unstable versions etc without breaking anything)

What do you reckon?

As I said at the top, this could be a followup PR if it makes sense to people! @ascjones I wonder what you reckon to this too?

Reply:
(this would mean adding a way to go from latest metadata to V14, V15 etc but going backwards should almost entirely mean just throwing away some information, and if it ever is harder than that then we can opt to break and remove an `into_runtime_metadata_vxx()` fn if we want)

Reply:
@cmichi I have this memory that you wanted some function or data added to the `Metadata` struct in Subxt a while ago (and also a memory of thinking I'd eventually forget that it was wanted since Subxt doesn't need it).

If we stopped exposing _all_ methods on `Metadata` but instead gave you a way to get back the `RuntimeMetadataV14` as described above, would that be sufficient? I'm quite eager to hide said APIs if I can, because they basically just exist for Subxt internal usage and that's crept out into the real world a bit too much, so I think the frame-metadata structs are a better thing to provide access to :)

Reply:
Just to follow up; Alex pointed out on our call this morning that a bunch of traits rely on being able to get information out of metadata (and the functions exposed on Metadata, while they could be refined, exist as more optimal ways to get at what we need).

so, while we could have a RuntimeMetadata new type to hide the underlying metadata version we use, we'd still need to provide enough information in the Metadata interface to get everything anybody might need for decoding/encoding things.

So this needs more thought, basically!